# 2025/07/06(日)
## 📚 プラクティス『プログラムの修正（リバーシ編）』


## 🧩 タスクばらし
- [x] プラクティスの内容を咀嚼
- [x] タスクばらし、スモールステップの作成
- [x] ソースコードの言語化
- [ ] ソースコードを修正
- [ ] 課題提出


## 🐾 スモールステップ
<details><summary>プラクティスの内容を咀嚼</summary>

- [x] プラクティスの内容を咀嚼
</details>

<details><summary>タスクばらし、スモールステップの作成</summary>

- [x] タスクばらし、スモールステップの作成
</details>

<details><summary>ソースコードの言語化</summary>

- [x] ソースコードの言語化 **※①から順に行う**
  - [x] ① reversi_methods_test.rb
  - [x] ② reversi.rb
  - [x] ③ reversi_methods.rb
  - [x] ④ position.rb


</details>

<details><summary>ソースコードを修正</summary>

- [ ] ソースコードを修正
  - test_cannot_put_stone
    - [ ] 修正
  - test_finished_of_quickest_win_board
    - [ ] 修正
  - test_put_stone
    - [ ] 修正
  - test_turn
    - [ ] 修正
  - test_finished_of_full_board
    - [ ] 修正

</details>

<details><summary>課題提出</summary>

- [ ] 修正したソースコードを Pull Request としてアップ
- [ ] Pull Request としてアップした URL とテストコードの実行結果を提出
</details>


---


## 🧑🏻‍💻 本日の取り組み
### グループコーチングの提出物の作成
- グループコーチング当日までブラッシュアップを行う

### プラクティス『プログラムの修正（リバーシ編）』
- チェリー本：第9章『例外処理を理解する』、第11章『パターンマッチを理解する』


---


## 🎯 次回
### グループコーチングの提出物の作成
- グループコーチング当日までブラッシュアップを行う

### プラクティス『プログラムの修正（リバーシ編）』
- `binding.break`を書き込み、挙動を一つずつ確認する
    

---


## 💡 本日の学び・気付き
### 復習：チェリー本 第9章『例外処理を理解する』
- 例外（Exception）とは、プログラムの実行中に発生した「例外的な問題」のこと。つまり「**エラーが起きてプログラムの実行を続けることができなくなった状態**」。
- 意図的にエラーを捕捉し、プログラムを実行させることもできる。また、捕捉するだけでなく、例外を**意図的に発生させる**こともできる。
- 例外を**適切に扱えば堅牢で読みやすいコードができあがる**し、万一エラーが起きた場合でも原因の調査がしやすくなる。例外を適切に扱わなかった場合はその反対。

#### 例外の捕捉
`1 + '10'`を irb で実行したときのエラーコードを見てみる。
```shell
irb(main):001> 1 + '10'
(irb):1:in `+': String can't be coerced into Integer (TypeError)
	from (irb):1:in `<main>'
	from <internal:kernel>:187:in `loop'
	from /Users/yoshiwo/.rbenv/versions/3.3.1/lib/ruby/gems/3.3.0/gems/irb-1.11.0/exe/irb:9:in `<top (required)>'
	from /Users/yoshiwo/.rbenv/versions/3.3.1/bin/irb:25:in `load'
	from /Users/yoshiwo/.rbenv/versions/3.3.1/bin/irb:25:in `<main>'
```
- `String can't be coerced into Integer`
  - 「String を Integer に変換することはできない」という意味。
- `(TypeError)`
  - 発生した例外のクラス名を示す。
  - Ruby では例外も**例外クラスのインスタンス**（例外オブジェクト）になっている。
- (irb):1:in `+'
  - 例外が発生した場所を示している。
- 3行目以降の情報
  - 例外が発生するまでの**メソッドの呼び出し履歴**。この情報を**バックトレース**や**スタックトレース**と呼ぶ。そして、このバックトレースは**デバッグする際の重要な情報源**になる。

#### 例外を捕捉して処理を続行する場合
もし、例外が発生してもプログラムを続行したい場合は、**例外処理を明示的に書く**ことでプログラムを続行させることが可能。以下は、その構文。
```ruby
begin
  # 例外が起きうる処理
rescue
  # 例外が発生した場合の処理
end
```

#### 例外オブジェクトから情報を取得する
Ruby では発生した例外自身もオブジェクトになっている。そのため例外オブジェクトのメソッドを呼び出すことで、発生した例外に関する情報を取得することができる。例外オブジェクトから情報を取得したい場合は次のような構文を使う。
```ruby
begin
  # 例外が起きうる処理
rescue => 例外オブジェクトを格納する変数
  # 例外が発生した場合の処理
end
```

### 復習：チェリー本 第11章『パターンマッチを理解する』
第11章は、パターンマッチの基本とさまざまな利用パターンを再読。


---


## ✍🏻 感想
### ✅ 復習終了
今日はチェリー本の復習だけでした。大きなトピックはありませんが、心身の健康を大切にしつつプラクティスを続けます。


---


### ⏰ 学習時間
- Today:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 4 hours 09 min
- Total: 1151 hours 07 min
